packages/vm/lib/runBlockchain.ts

```diff
@@ -1,100 +1,40 @@
+      parentState = parentBlock.header.stateRoot
+      // generate genesis state if we are at the genesis block
+      // we don't have the genesis state
+      if (!headBlock) {
+        await this.stateManager.generateCanonicalGenesis()
       } else {
-        resolve()
+        parentState = headBlock.header.stateRoot
       }
-    })
-
-    function processBlock(block: any, reorg: boolean, cb: any) {
-      async function _processBlock() {
-        // determine starting state for block run
-        function getStartingState(cb: any) {
-          // if we are just starting or if a chain re-org has happened
-          if (!headBlock || reorg) {
-            blockchain.getBlock(block.header.parentHash, function (err: any, parentBlock: any) {
-              parentState = parentBlock.header.stateRoot
-              // generate genesis state if we are at the genesis block
-              // we don't have the genesis state
-              if (!headBlock) {
-                const generateCanonicalGenesis = callbackify(
-                  self.stateManager.generateCanonicalGenesis.bind(self.stateManager),
-                )
-                return generateCanonicalGenesis(cb)
-              } else {
-                cb(err)
-              }
-            })
-          } else {
-            parentState = headBlock.header.stateRoot
-            cb()
-          }
-        }
-
-        // run block, update head if valid
-        function runBlock(cb: any) {
-          self
-            .runBlock({
-              block: block,
-              root: parentState,
-            })
-            .then(() => {
-              // set as new head block
-              headBlock = block
-              cb()
-            })
-            .catch((err) => {
-              // remove invalid block
-              blockchain.delBlock(block.header.hash(), function () {
-                cb(err)
-              })
-            })
-        }
+    }
 
-        await new Promise((resolve, reject) =>
-          getStartingState(function (error: any) {
-            if (error) {
-              reject(error)
-            } else {
-              resolve()
-            }
-          }),
-        )
-          .then(function () {
-            return new Promise((resolve) =>
-              runBlock(function (error: any) {
-                if (error) {
-                  reject(error)
-                } else {
-                  resolve()
-                }
-              }),
-            )
-          })
-          .then(function () {
-            cb()
-          })
-          .catch(function (error: any) {
-            cb(error)
-          })
-      }
-      _processBlock()
+    // run block, update head if valid
+    try {
+      await this.runBlock({ block, root: parentState })
+      // set as new head block
+      headBlock = block
+    } catch (error) {
+      // remove invalid block
+      await blockchain.delBlock(block.header.hash())
+      throw error
```

Comment:
WHuuaaH! We will soon be in a state that normal people will be able to read our code.

I am so excited! 😄  😛 😄

Reply:
But honestly: this "before" -> "after" is really insane. Sitting here - sighting.